src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -235,70 +244,358 @@ internal static void FailFast(string message, Exception? exception, RhFailFastRe
+        private static unsafe bool WriteException(Exception exception, int maxMessageSize, int maxNumberStackFrames, int maxMethodNameSize)
+        {
+            if (!WriteHexValue("address", (ulong)Unsafe.AsPointer(ref exception)))
+                return false;
+
+            if (!WriteHexValue("hr", exception.HResult))
+                return false;
+
+            if (!WriteValue("message", exception.Message, maxMessageSize))
+                return false;
+
+            // Exception type names are not truncated because the full name is important to bucketing and usually not that long
+            if (!WriteValue("type", exception.GetType().ToString()))
+                return false;
+
+            bool success = WriteBlock("stack", '[', ']', () =>
+            {
+                StackTrace stackTrace = new(exception);
+                int count = 0;
+                foreach (StackFrame frame in stackTrace.GetFrames())
+                {
+                    // Check if the stack frame limit has been hit
+                    if (++count > maxNumberStackFrames)
+                        break;
+
+                    // Write as many stack frames that will fit
+                    if (!WriteStackFrame(frame, maxMethodNameSize))
+                        break;
+                }
+                return true;
+            });
+
+            if (!success)
+                return false;
+
+            AggregateException? aggregate = exception as AggregateException;
+            if (aggregate is not null || exception.InnerException is not null)
+            {
+                success = WriteBlock("inner", '[', ']', () =>
+                {
+                    // Write as many inner exceptions that will fit
+                    if (aggregate is not null)
+                    {
+                        foreach (Exception ex in aggregate.InnerExceptions)
+                        {
+                            if (!WriteBlock(null, '{', '}', () => WriteException(ex, maxMessageSize, maxNumberStackFrames, maxMethodNameSize)))
+                                break;
+                        }
+                    }
+                    else
+                    {
+                        WriteBlock(null, '{', '}', () => WriteException(exception.InnerException, maxMessageSize, maxNumberStackFrames, maxMethodNameSize));
+                    }
+                    return true;
+                });
+
+                if (!success)
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Write an exception stack frame to the triage buffer
+        /// </summary>
+        /// <param name="frame">the stack frame instance to write</param>
+        /// <param name="maxNameSize">limits the size of the frame type name</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private static bool WriteStackFrame(StackFrame frame, int maxNameSize)
```

Comment:
The mini json writer should be its own type in its own .cs file. These Write methods should be instance methods on the struct.

Reply:
I extracted all the JSON writing and the higher level crash info rendering all in the CrashInfo type.  I think there might be a better name for it and it is abstracts how the triage info is written not just the JSON writing.